File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[] lis = new int[n];
        int res = 0;
        
        Arrays.fill(lis, 1);
        
        for(int i = n - 2 ; i >= 0 ; i--) {
            for(int j = i + 1; j < n; j++) {
                if (nums[i] < nums[j]) {
                    lis[i] = Math.max(lis[i], lis[j] + 1);
                }
            }
        }
        
        for(int i = 0; i < n; i++) {
            if (lis[i] > res) {
                res = lis[i];
            }
        }
        
        return res;
        
    }
}

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size(), res = 0;
        vector<int> lis(n, 1); // lis[i] represents the the length of Longest Increasing Subsequence that starts at i, the minum guarenteed value is 1
        for (int i = n - 2; i >= 0; i--) {
            for (int j = i + 1; j < n; j++) {
                if (nums[i] < nums[j]) lis[i] = max(lis[i], lis[j] + 1);
            }
        }
        
        for (int val : lis) {
            if (val > res) res = val;
        }
        return res;
    }
};